src/Http/Owin/src/OwinEnvironment.cs

```diff
@@ -478,4 +399,213 @@ public FeatureMap(Func<TFeature, object> getter, Func<object> defaultFactory, Ac
+
+            { OwinConstants.SendFiles.SendAsync, new FeatureMap<IHttpResponseBodyFeature>(feature => new SendFileFunc(feature.SendFileAsync)) },
+            { OwinConstants.Security.User, new FeatureMap<IHttpAuthenticationFeature>(feature => feature.User, ()=> null, (feature, value) => feature.User = Utilities.MakeClaimsPrincipal((IPrincipal)value), () => new HttpAuthenticationFeature()) },
+            { OwinConstants.RequestId, new FeatureMap<IHttpRequestIdentifierFeature>(feature => feature.TraceIdentifier, ()=> null, (feature, value) => feature.TraceIdentifier = (string)value, () => new HttpRequestIdentifierFeature()) },
+            { OwinConstants.CallCancelled, new FeatureMap<IHttpRequestLifetimeFeature>(feature => feature.RequestAborted) },
+
+            { OwinConstants.CommonKeys.ClientCertificate, new FeatureMap<ITlsConnectionFeature>(feature => feature.ClientCertificate, (feature, value) => feature.ClientCertificate = (X509Certificate2)value) },
+            { OwinConstants.CommonKeys.LoadClientCertAsync, new FeatureMap<ITlsConnectionFeature>(feature => new Func<Task>(() => feature.GetClientCertificateAsync(CancellationToken.None))) },
+            { OwinConstants.WebSocket.AcceptAlt, new FeatureMap<IHttpWebSocketFeature>(feature => new WebSocketAcceptAlt(feature.AcceptAsync)) }
+        };
+
+        /// <remarks>
+        /// Will be used, only if <see cref="FeatureMaps"/> or <see cref="Clear"/> is called from user-code.
+        /// Is a deep-copy of the singleton <see cref="_entries"/>
+        /// </remarks>
+        private IDictionary<string, FeatureMap> _contextEntries;
+
+        /// <remarks>
+        /// In order not to copy the whole dictionary of featureMaps per request,
+        /// and since OWIN allows the <see cref="Remove(string)"/> operation,
+        /// it's more lightweight to keep track of deleted keys.
+        /// </remarks>
+        private HashSet<string> _deletedKeys;
+
+        /// <remarks>
+        /// There are some <see cref="FeatureMap"/> entries that are context-dependent.
+        /// This dictionary is allocated per request, but does not contain as many entries as <see cref="_entries"/>.
+        /// </remarks>
+        private readonly IDictionary<string, FeatureMap> _contextDependentEntries;
+
+        public OwinEntries(HttpContext context)
+        {
+            _contextDependentEntries = new Dictionary<string, FeatureMap>
+            {
+                { OwinConstants.ResponseBody, new FeatureMap<IHttpResponseBodyFeature>(feature => feature.Stream, () => Stream.Null, (feature, value) => context.Response.Body = (Stream)value) }, // DefaultHttpResponse.Body.Set has built in logic to handle replacing the feature.
+            };
+        }
+
+        static readonly string[] PortStrings = CreatePortStrings();
+        static string[] CreatePortStrings()
+        {
+            var ports = new string[65535]; // limit of ephemeral ports https://en.wikipedia.org/wiki/Ephemeral_port
```

Comment:
Do we really want to preallocate so many remote/ephemeral port strings? This reminds me of Java's integer pool but way bigger! 😆

I know this array is only about a meg strings included, and it shouldn't take too long to initialize at startup, but I imagine the vast majority apps will never read the remote port, considering it's only really useful at the application layer as a pseudo connection id. Is the remote port read more than I realize?

Reply:
> Do we really want to preallocate so many remote/ephemeral port strings?
as per benchmarks results, I was also a bit concerned here. Maybe we should have pre-allocated known ports like `443`, `80`, but not the whole pool. It seems like for the "fixed" version all of the allocations are on 13 MB because of that specific reason - we just allocate a lot of strings and we dont see actual diff here.

> but I imagine the vast majority apps will never read the remote port
agree

> Is the remote port read more than I realize?
I can only confirm that coming from the internal team implementing proxy, we were interested in getting that telemetry. And that's why we were reading both local and remote ports. I dont know if this is being used on a hot-path by many people who are still using Owin. 

Let me do the following - I will pre-allocate known ports, and for others we will do the `.ToString()` call. I will update benchmarks once i have results

Reply:
Removed the allocated buffer of port string representations and updated benchmark results.

Reply:
Yes, this case was pretty specific to the internal usage - we need to record both local and remote ports in telemetry and that had a reasonable allocation cost at high traffic throughput.